transports/quic/src/transport.rs

```diff
@@ -0,0 +1,641 @@
+                    socket_addr.is_ipv6(),
+                )?);
+            }
+            dialer
+                .as_mut()
+                .unwrap()
+                .pending_dials
+                .push_back(to_endpoint);
+        } else {
+            // Pick a random listener to use for dialing.
+            let n = rand::random::<usize>() % listeners.len();
+            let listener = listeners.get_mut(n).expect("Can not be out of bound.");
+            listener.pending_dials.push_back(to_endpoint);
+            if let Some(waker) = listener.waker.take() {
+                waker.wake()
+            }
+        };
+
+        Ok(async move {
+            let connection = rx
+                .await
+                .map_err(|_| TransportError::EndpointDriverCrashed)?
+                .map_err(TransportError::Reach)?;
+            let final_connec = Upgrade::from_connection(connection).await?;
+            Ok(final_connec)
+        }
+        .boxed())
+    }
+
+    fn dial_as_listener(
+        &mut self,
+        addr: Multiaddr,
+    ) -> Result<Self::Dial, CoreTransportError<Self::Error>> {
+        // TODO: As the listener of a QUIC hole punch, we need to send a random UDP packet to the
+        // `addr`. See DCUtR specification below.
+        //
+        // https://github.com/libp2p/specs/blob/master/relay/DCUtR.md#the-protocol
+        self.dial(addr)
+    }
+
+    fn poll(
+        mut self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+    ) -> Poll<TransportEvent<Self::ListenerUpgrade, Self::Error>> {
+        if let Some(dialer) = self.ipv4_dialer.as_mut() {
+            if dialer.drive_dials(cx).is_err() {
+                // Background task of dialer crashed.
+                // Drop dialer and all pending dials so that the connection receiver is notified.
+                self.ipv4_dialer = None;
+            }
+        }
+        if let Some(dialer) = self.ipv6_dialer.as_mut() {
+            if dialer.drive_dials(cx).is_err() {
+                // Background task of dialer crashed.
+                // Drop dialer and all pending dials so that the connection receiver is notified.
+                self.ipv4_dialer = None;
+            }
+        }
+        match self.listeners.poll_next_unpin(cx) {
+            Poll::Ready(Some(ev)) => Poll::Ready(ev),
+            _ => Poll::Pending,
+        }
+    }
+}
+
+#[derive(Debug)]
+struct Dialer {
+    endpoint_channel: EndpointChannel,
+    pending_dials: VecDeque<ToEndpoint>,
+}
+
+impl Dialer {
+    fn new<P: Provider>(
+        config: Config,
+        is_ipv6: bool,
+    ) -> Result<Self, CoreTransportError<TransportError>> {
+        let endpoint_channel =
+            EndpointChannel::new_dialer::<P>(config, is_ipv6).map_err(CoreTransportError::Other)?;
+        Ok(Dialer {
+            endpoint_channel,
+            pending_dials: VecDeque::new(),
+        })
+    }
+
+    fn drive_dials(&mut self, cx: &mut Context<'_>) -> Result<(), mpsc::SendError> {
```

Comment:
Should this be called `poll`?

Reply:
This would force us to return `Poll::Pending` if we don't have any `pending_dials`. This thought makes me think, shouldn't this struct contain a waker that we call when we are pushing a new pending dial?

It might make sense to move this struct into its own module so we can properly encapsulate these fields and `QuicTransport` can't just modify them at will :)

Reply:
> This thought makes me think, shouldn't this struct contain a waker that we call when
 we are pushing a new pending dial?

I think I originally only added a waker for the `Listener` and not for the `Dialer` be
cause I thought that after initiating a new dial a user would continue polling the swa
rm / transport anyway. But thinking about it now there are cases were this assumption
is incorrect. Added a waker in eafa89e.

> Should this be called poll?
>
> This would force us to return Poll::Pending if we don't have any pending_dials.

Not sure if that's intuitive. `Poll::Pending` usually means "there is work to do but w
e are currently blocked", which is not the case if the `pending_dials` are empty.

Reply:
> > Should this be called poll?
> > This would force us to return Poll::Pending if we don't have any pending_dials.
> 
> Not sure if that's intuitive. `Poll::Pending` usually means "there is work to do but w e are currently blocked", which is not the case if the `pending_dials` are empty.

I don't quite see it that way. To me `Poll::Pending` means "I cannot proceed with the task that you want me to do". Reading from an empty queue is a legit case to register a waker IMO. It is the same thing for channels. Reading from an empty channel does not return `Ok(())` but `Poll::Pending`.

Reply:
Resolved with 7a86e30.